flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/AsyncExecutionController.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.asyncprocessing;
+
+import org.apache.flink.api.common.operators.MailboxExecutor;
+import org.apache.flink.api.common.state.v2.State;
+import org.apache.flink.core.state.InternalStateFuture;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+/**
+ * The Async Execution Controller (AEC) receives processing requests from operators, and put them
+ * into execution according to some strategies.
+ *
+ * <p>It is responsible for:
+ * <li>Preserving the sequence of elements bearing the same key by delaying subsequent requests
+ *     until the processing of preceding ones is finalized.
+ * <li>Tracking the in-flight data(records) and blocking the input if too much data in flight
+ *     (back-pressure). It invokes {@link MailboxExecutor#yield()} to pause current operations,
+ *     allowing for the execution of callbacks (mails in Mailbox).
+ *
+ * @param <R> the type of the record
+ * @param <K> the type of the key
+ */
+public class AsyncExecutionController<R, K> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AsyncExecutionController.class);
+
+    public static final int DEFAULT_MAX_IN_FLIGHT_RECORD_NUM = 6000;
+
+    /** The max allow number of in-flight records. */
+    private final int maxInFlightRecordNum;
+
+    /** The key accounting unit which is used to detect the key conflict. */
+    final KeyAccountingUnit<R, K> keyAccountingUnit;
+
+    /**
+     * A factory to build {@link org.apache.flink.core.state.InternalStateFuture}, this will auto
+     * wire the created future with mailbox executor. Also conducting the context switch.
+     */
+    private final StateFutureFactory<R, K> stateFutureFactory;
+
+    /** The state executor where the {@link StateRequest} is actually executed. */
+    final StateExecutor stateExecutor;
+
+    /** The corresponding context that currently runs in task thread. */
+    RecordContext<R, K> currentContext;
```

Comment:
The `currentContext` is fully managed by `AEC` and related runtime components, while the `handleRequest` is called by the direct implementation of `State APIs`, which is somewhat upper layer of `AEC` and lacks the knowledge of this context.

Reply:
I am not quite sure about the "fully managed". In my opinion, "fully managed" is equivalent that the context is created/used/destroyed by AEC. Could you give an example -- whose may invoke buildContext() to create a `RecordContext`?

Reply:
You're right about "fully managed", but I'm saying it is fully managed by `AEC` AND **related runtime components**. That means this is managed by processing framework instead of the caller of `handleRequest`, which is the `State`-related implementation.

The `buildContext` will be invoked right before the `processElement`. To be more specific, my brief idea is, both `buildContext` and `processElement` will be packed into one lambda in `RecordProcessorUtils#getRecordProcessor`.

Reply:
Thank you for your apply. I think I get a more clear view on the `RecordContext` now. From the testing code, I find the lifecycle of `RecordContext` seems to be 
```
buildContext() (named as ctx1) -> aec.setCurrentContext(ctx1) -> user_code -> 

async state -> aec.handleRequest() -> new StateRequest(ctx1) -> async state -> 

aec.handleRequest() -> new StateRequest(ctx1)
```

This async state operations form a chain, and all operations in the chain use the `ctx1`.